### ISSUE TITLE
XQuartz and Anaconda Python building and library linking errors

### ISSUE THREAD

**Issue by kylemcdonald** [davisking/dlib#437]
tl;dr: I fixed errors building dlib by disabling the GUI, and fixed errors linking to libmkl_rt.dylib by modifying the .so with `install_name_change`.

I'm using OSX 10.11 with XCode 8.2.1 and Anaconda. I have CUDA 8.0.55 and CuDNN 5.1.10.

```
$ brew cask install Caskroom/cask/xquartz
$ brew install boost boost-python
$ mkdir build; cd build; cmake .. ; cmake --build .
```

CUDA and CuDNN are successfully detected + built by `cmake ..`. But `cmake --build .` gives me an error pretty early on:

```
[  1%] Building CXX object dlib/CMakeFiles/dlib.dir/gui_widgets/fonts.cpp.o
In file included from /Users/kyle/dlib/dlib/gui_widgets/fonts.cpp:14:
/Users/kyle/dlib/dlib/gui_widgets/nativefont.h:313:21: error: use of undeclared identifier 'XAllocColor'
```

Instead of digging into this, I just disabled the GUI:

```
$ cmake .. -DDLIB_NO_GUI_SUPPORT:BOOL=ON
$ cmake --build .
```

And it build dlib successfully. Time to build Python support (specifying no GUI here too):

```
$ cd .. ; python setup.py install --yes DLIB_NO_GUI_SUPPORT
```

Success! But then...

```
$ python
kyle@kyle ~/D/L/dlib (master)> python
Python 2.7.12 |Anaconda 4.0.0 (x86_64)| (default, Jul  2 2016, 17:43:17) 
[GCC 4.2.1 (Based on Apple Inc. build 5658) (LLVM build 2336.11.00)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
Anaconda is brought to you by Continuum Analytics.
Please check out: http://continuum.io/thanks and https://anaconda.org
>>> import dlib
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/kyle/anaconda/lib/python2.7/site-packages/dlib-19.2.99-py2.7-macosx-10.6-x86_64.egg/dlib/__init__.py", line 1, in <module>
    from .dlib import *
ImportError: dlopen(/Users/kyle/anaconda/lib/python2.7/site-packages/dlib-19.2.99-py2.7-macosx-10.6-x86_64.egg/dlib/dlib.so, 2): Library not loaded: libmkl_rt.dylib
  Referenced from: /Users/kyle/anaconda/lib/python2.7/site-packages/dlib-19.2.99-py2.7-macosx-10.6-x86_64.egg/dlib/dlib.so
  Reason: image not found
```

Tried to learn more about this:

```
$ otool -L ~/anaconda/lib/python2.7/site-packages/dlib-19.2.99-py2.7-macosx-10.6-x86_64.egg/dlib/dlib.so | grep mkl
	libmkl_rt.dylib (compatibility version 0.0.0, current version 0.0.0)
```

To me this looks like it should have an `@rpath` at the beginning? So I do this:

```
$ install_name_change -change libmkl_rt.dylib @rpath/libmkl_rt.dylib ~/anaconda/lib/python2.7/site-packages/dlib-19.2.99-py2.7-macosx-10.6-x86_64.egg/dlib/dlib.so
$ python
>> import dlib
>>
```

Success! Hopefully... maybe I just delayed the inevitable, and it was linked to another dylib? Then I go to run the Python examples:

```
$ cd ../python_examples ; python face_recognition.py
Traceback (most recent call last):
  File "face_recognition.py", line 50, in <module>
    import dlib
ImportError: dlopen(/Users/kyle/Documents/Learning/dlib/python_examples/dlib.so, 2): Library not loaded: libmkl_rt.dylib
  Referenced from: /Users/kyle/Documents/Learning/dlib/python_examples/dlib.so
```

Looks like there's another copy of `dlib.so` here. I suppose I could just delete it, or make the same change. I opt to make the same change:

```
$ install_name_change -change libmkl_rt.dylib @rpath/libmkl_rt.dylib dlib.so
$ python face_recognition.py 
Call this program like this:
   ./face_recognition.py shape_predictor_68_face_landmarks.dat dlib_face_recognition_resnet_model_v1.dat ../examples/faces
You can download a trained facial shape predictor and recognition model from:
    http://dlib.net/files/shape_predictor_68_face_landmarks.dat.bz2
    http://dlib.net/files/dlib_face_recognition_resnet_model_v1.dat.bz2
```

Getting closer...

```
$ curl -O http://dlib.net/files/shape_predictor_68_face_landmarks.dat.bz2
$ curl -O http://dlib.net/files/dlib_face_recognition_resnet_model_v1.dat.bz2
$ bunzip2 shape_predictor_68_face_landmarks.dat.bz2
$ bunzip2 dlib_face_recognition_resnet_model_v1.dat.bz2
$ python face_recognition.py shape_predictor_68_face_landmarks.dat dlib_face_recognition_resnet_model_v1.dat ../examples/faces
Traceback (most recent call last):
  File "face_recognition.py", line 74, in <module>
    win = dlib.image_window()
AttributeError: 'module' object has no attribute 'image_window'
```

Right, I didn't install the GUI components. I comment out all the lines referencing the `win` variable, and:

```
$ python face_recognition.py shape_predictor_68_face_landmarks.dat dlib_face_recognition_resnet_model_v1.dat ../examples/faces
Processing file: ../examples/faces/2007_007763.jpg
Number of faces detected: 7
Detection 0: Left: 93 Top: 194 Right: 129 Bottom: 230
-0.0512203
0.0150111
0.0642802
-0.0438789
...
```

Success!

Overall I'm posting this issue to ask a couple questions:

1. Should I have done something differently with XQuartz at the beginning?
2. Is there a variable I should set in CMake to work with Anaconda correctly? I saw in another issue you recommend just disabling Anaconda completely for building and running dlib. Is this still the recommendation, or is there another way?

**Comment by davisking** [comment]
XQuartz is fine.  The reason you get this error is because anaconda, for some reason I'll never understand, includes copies of the X11 header files with a bunch of X11 functions removed.  So if your compiler's include search path specifies to look in the anaconda location first then it picks up these bad X11 header files and BOOM, errors about missing X11 functions.  This is especially irritating since the X11 API is ancient, maybe older than me, I forget the exact history.  So I don't see any reasonable excuse for why anaconda would include broken X11 headers.  I usually tell people to delete anaconda at this point.  But maybe you don't want to do that.  

A better and more permanent solution would be to modify the CMake scripts to adjust the compiler include search path to somehow avoid anaconda's X11 files.  You can see where the X11 path is located in dlib/CMakeLists.txt here: https://github.com/davisking/dlib/blob/master/dlib/CMakeLists.txt#L252.  You can even see that there is a comment there that the current way it does it should avoid anaconda's nonsense, but apparently it doesn't anymore :)

Also, in general, you don't need to set rpath to use the mkl.  This kind of thing is generally handled by some system wide setting like adding the MLK folder to your LD_LIBRARY_PATH or some other system list of folders the runtime linker searches for shared libraries.  I don't recall the setting on the mac.  But on most unix systems it's setting LD_LIBRARY_PATH.  See also: https://en.wikipedia.org/wiki/Rpath.  So this isn't a dlib issue, just a matter of installing the MKL correctly. 

To summarize, you can use dlib with anaconda.  You just need to make sure the anaconda X11 header files aren't found since they are broken.  If you find an adjustment to cmake that avoids them please do submit a PR :)

**Comment by davisking** [comment]
Also, you can delete the copy of dlib.so in the python_examples folder.  It's automatically copied there by the cmake build scripts.  This is done so that users don't have to install dlib in some python folder to run the examples.  You just build it and it's in the examples fold and the examples are runnable.  But yeah, if you are looking to install it then it's superfluous.

**Comment by kylemcdonald** [comment]
It looks like what's happening right now with XQuartz and Anaconda both installed is that `X11_FOUND` is true, but the variables get set to `X11_INCLUDE_DIR=~/anaconda/include/opt/X11/include` and `X11_LIBRARIES=/opt/X11/lib/libSM.dylib/opt/X11/lib/libICE.dylib/opt/X11/lib/libX11.dylib/opt/X11/lib/libXext.dylib`. I'm not great with cmake but I'll try to find and submit a fix for this case.

**Comment by davisking** [comment]
Yeah, I looked a little to see if there is a way to tell cmake to exclude a
folder from the search path and didn't find anything.  I'm sure there is
some way to solve this in cmake though.  Maybe the simplest thing is to add
whatever folder has the correct path to the HINTS or PATHS for find_path so
it's found first.

**Comment by davisking** [comment]
Or yeah, maybe editing the X11_INCLUDE_DIR will work.  Let me know :)

**Comment by davisking** [comment]
Guess I'm too slow to comment here, already fixed :)

**Comment by kylemcdonald** [comment]
Also, you're right that `Library not loaded: libmkl_rt.dylib` can be fixed by adding a line to my profile. I use the Fish shell, but in Bash I think it would be:

```
$ echo "export DYLD_LIBRARY_PATH=$DYLD_LIBRARY_PATH:/Users/your_username!/anaconda/lib" >> ~/.bash_profile
$ source ~/.bash_profile
```

Thanks!

**Comment by davisking** [comment]
I'm not sure how it works on OS X, but in linux you can edit /etc/ld.so.conf to add folders to the library search path for all users all the time.  That's normally how libraries are installed in linux.  Then it works no matter how your environment variables are set.  I'm sure OS X must have a similar mechanism since you don't have to add the many other shared libraries needed to run any application to your LD_LIBRARY_PATH.

**Comment by anubhavshrimal** [comment]
The update didn't work on my mac.
As an alternative, I simply deleted the X11 folder in the `anaconda/include` folder before running `pip install dlib`.
It worked out for me. 
Then I added the folder back into the directory.

**Comment by Hi-king** [comment]
If you use anaconda, now you can install dllib from menpo.

`conda install -c menpo dlib`
